cmd/lb/main.go

```diff
@@ -0,0 +1,252 @@
+
+	if err := app.Run(os.Args); err != nil {
+		logger.Fatal(err)
+	}
+}
+
+func createApp() *cli.App {
+	app := &cli.App{
+		Action: run,
+		Flags:  flags,
+	}
+	return app
+}
+
+func run(c *cli.Context) error {
+	app := fx.New(runtime.NewServer(c.Context, c),
+		fx.Populate(&logger, &n),
+		fx.Supply(c))
+
+	if err := app.Start(c.Context); err != nil {
+		return err
+	}
+	defer app.Stop(context.Background())
+
+	logger.Info("server started")
+
+	if c.Bool("gateway") {
+		return runGateway(c, n)
+	}
+
+	return runWorker(c, n)
+}
+
+func runGateway(c *cli.Context, n *server.Node) error {
+	log := logger.With(n)
+	log.Info("gateway started")
+
+	ch := csp.NewChan(&csp.SyncChan{})
+	defer ch.Close(c.Context)
+
+	n.Vat.Export(chanCap, chanProvider{ch})
+
+	log.Info("exported channel")
+
+	return http.ListenAndServe(":8080", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// each HTTP request is served in its own goroutine
+
+		// Get the request payload
+		b, err := io.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		// Block on <-ch until either we get a result, or the request
+		// context expires.  In the latter case, this indicates that
+		// the client aborted.
+		f, release := ch.Recv(r.Context())
+		defer release()
+
+		// Get the worker's echo capability
+		echo := lb.Echo(f.Client())
+
+		// Call the handler's RPC method
+		e, release := echo.Echo(r.Context(), lb.Data(b))
+		defer release()
+
+		// block until we get the RPC response
+		res, err := e.Struct()
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadGateway)
+			return
+		}
+
+		// Grab the echoed bytes from the response
+		p, err := res.Response()
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadGateway)
+			return
+		}
+
+		// Send the response back to the caller
+		_, err = io.Copy(w, bytes.NewReader(p))
+		if err != nil {
+			log.WithError(err).
+				Warn("failed to write echo response")
+		}
+	}))
+}
+
+func runWorker(c *cli.Context, n *server.Node) error {
+	if err := waitGateway(c, n); err != nil {
+		return err
+	}
+
+	// Routing information for the gateway server.
+	gateway, err := peer.Decode(c.String("dial"))
+	if err != nil {
+		return err
+	}
+
+	addr := peer.AddrInfo{ID: gateway}
+
+	log := logger.With(&addr)
+	log.Info("worker started")
+
+	// Establish connection with gateway (corresponding to channel capability that gateway exported earlier)
+	conn, err := n.Vat.Connect(c.Context, addr, chanCap)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	// Recover channel capability from Gateway.  Cast it as a send-
+	// only channel, to avoid mistakes.
+	ch := csp.Sender(conn.Bootstrap(c.Context))
+
+	// Loop until the worker starts shutting down.
+	for c.Context.Done() == nil {
+		// Setup of the echo capability. Start an echo server, and derive a client from it.
+		server := lb.EchoServer{}
+		e := capnp.Client(lb.Echo_ServerToClient(server))
```

Comment:
When I put this outside the for loop each worker would serve a request, and then each subsequent one would have capnp error. Is this because the gateway releases the capability each time? like is there a way to redesign this to reuse the capabilities so they don't have to be generated/discarded after each use or is it not a big deal